### DISCUSSION TITLE
CPU overflow on Airflow subprocesses

### DISCUSSION THREAD

**Discussion by 4andy** [apache/airflow#32871] [Q&A] [answered]
Hi, We have been running a few Airflow instances for the past few years and they have been very stable. We have very lightweight needs. We have upgraded airflow a few times and they have continued to be stable until this issue. Over the past couple of months, all of the instances have gone into a cpu overflow condition on various Airflow subprocesses; however, the issue has occurred at different times on each server seemingly randomly and we have not been able to identify a trigger for the condition. Any ideas on further troubleshooting would be appreciated.

### Environment
All servers - Red Hat Enterprise Linux Server 7.9, LocalExecutor
Airflow: 2.2.5, postgresql 12.2
Airflow: 2.5.1, postgresql 15.2

### Issue
Airflow subprocesses: (all: airflow scheduler - DagFileProcessor, airflow task runner, airflow task supervisor) begin to consume a full core of the machine or as much as they can get. There are no errors, tasks still complete successfully but they get slower and slower until the machine is overtaken. A task that generally takes 10 sec to run begins to run slower until it takes several minutes to run. Machine CPU usage climbs similarly. Memory usage on the server is flat and unchanging. 

On a healthy instance, the airflow subprocesses consume only a percent or two of a core. (We run at all default scheduler parameter settings). The issue cannot be resolved by stopping and restarting the airflow instance. However, if the linux server is restarted, the issue is resolved for days or weeks before it arises again.

### Things observed
* I have tried changing many of the scheduler parameters and they seem to have no effect on the issue. We only have ~ 50 dags with a few tasks each.
* Changing the _parsing_processes_ from the default of 2 to many more or less seems to have no effect. E.g. 8 DagFileProcessor process will all each pull a full core of the machine.
* Core airflow processes seem to be unaffected. airflow scheduler -- DagFileProcessorManager continues to run consuming only a percent or two of a core. The webserver is fine. 

The fact that restarting the linux server resolves the issue but restarting the airflow instance does not is a puzzlement for me and I have run out of ideas. As such, I thought I would reach out in hopes that someone may have additional insight or ideas that may help find the cause of the issue.

Thanks for any ideas or input you may have.

**Comment by potiuk** [comment]
IT's really hard to say but maybe a good idea would be use some profilers and see what Airflow is doing when it goes to "overdrive". Tools like https://github.com/benfred/py-spy can connect to running python interpreter and dump information on what's going on. 

Maybe you could try it and provide more information? 

Without it, it is shooting in the dark.

**Comment by 4andy** [comment]
Thanks for the suggestion @potiuk. We will give it a try and let you know what we find.

**Comment by 4andy** [comment]
One thing I did just notice - when the airflow instance is in the unhealthy state, the DagFileProcessor processes are short lived (perhaps lasting just a few seconds) and constantly being recreated (their pids keep changing).
![Screenshot 2023-08-07 120001](https://github.com/apache/airflow/assets/41160084/2b736569-6c2c-42c0-9592-592e800ac1df) 

In a healthy airflow instance, the DagFileProcessor processes are relatively long lived and consume a reasonable amount of cpu.
![Screenshot 2023-08-07 120009](https://github.com/apache/airflow/assets/41160084/f2ee9f9e-e9f4-44b7-a1d8-2cb675ef65b7)

When unhealthy, is it possible the DagFileProcessor processes are crashing and constantly being restarted?

**Reply by potiuk** [reply]
Are you using standalone DagFileProcessor or not? It's not clear from the description and your screenshots are only showing scheduler not `dag-processor` that I would expect to see. So I assume you do not have a standalone one and you know this is a dag fileprocessor from process name ? We cannot see the nae of the process in the screenshots :) 

I think there are scenarios that **some** conditions triggers the dag processor subprocess might fail quickly. The way how the processor works is that there is just ONE process running overloooking the rest  - named "airflow scheduler -- DagFileProcessorManager") and it forks a separate process for each of the files it parses - named "airflow scheduler - DagFileProcessor {file_path}". 

This is in order to isolate any parsing exceptions or problems from the main process that monitors the rest). 

So there are two options there:

1) the main dag file processor "airflow scheduler -- DagFileProcessorManager" -> this one should stay up for a long time. If it keeps on restarting, yes, this is a very probable root cause of the problem - we need to just find out why. 

2) any of the "forked" processes - "airflow scheduler - DagFileProcessor {file_path}" - those are only runnig during parsing of the individual files. For them, it is perfectly fine to live for a short period of time - they will just parse and serialize the DAG file, and once done, they exit.

I think if 1) is restarting then the root cause or at least some indication that things are going wrong shoudl be in scheduiler or processor logs. 

I think if 2) are crashing it's the same - unless there is something preventing them from writing the logs. The one thing that comes to my mind that the processes are runnig out of available file descriptors. That **could** result in the problems you might observe (because they would not be able to open and parse DAG files and at the same time it might prevent them from writing any information to the log files (because it might be impossible to open a file descriptor to write to the log file).  

At the same time it could work as you observe - as the amount of sockets being available is close to exhaution (i.e. there is a file descripto leak)  the continuous stopping and restarting procesess might occasionally free some of the sockets thus making the system works slower and slower because the pool of such remaining sockets after closing process might get smaller and smaller.

That's only a hypothesis of course. No idea if that's the reason but maybe you can check it ? 

 No idea what could be the root cause, but maybe if we can find out that something like that happens, it might bring us closer to finding a root cause.

**Reply by 4andy** [reply]
Thanks for the ideas - and yes I think we are in scenario 2 above and not using a standalone DagFileProcessor. DagFileProcessorManager is not crashing and long lived. Here are better screenshots. In the unhealthy state the DagFileProcessor processes are always present and maxed out cpu-wise. They last for a few seconds and are immediately recreated. 
![Screenshot 2023-08-08 092528](https://github.com/apache/airflow/assets/41160084/73f84e3b-9d10-49fc-9118-e5abb8f2a9b4)

In a healthy state, the DagFileProcessor processes just pop up now and then with the expected refresh schedule I think - I had to wait a bit to get this screen shot when one popped up. They don't last very long either but they don't consume much cpu. 
![Screenshot 2023-08-08 092518](https://github.com/apache/airflow/assets/41160084/4e99bd87-5479-4382-9c77-6045daf4d5b0)

We are looking into the file descriptor/socket possibility.

**Reply by 4andy** [reply]
Additionally, when in the unhealthy state, the airflow task runners and supervisor consume high CPU. Does anyone have insight into what these processes are doing and what may cause them to consume high cpu? When healthy, these processes consume very little cpu.
![Screenshot 2023-08-09 101956](https://github.com/apache/airflow/assets/41160084/a6b76548-8c47-45af-a160-2ebff78d98fa)

**Reply by potiuk** [reply]
Thos are celery processes.

And the sheer fact that you have it happening to those as well - especially the sipervisor indicates that there is a system-level problem not airflow process. You also mentioned that restarting airflow does not help (only restarting server)  - that indicates that the theory around some system-wide resources being exhausted - like sockets is very likely to be the root cause.

**Reply by potiuk** [reply]
Another possibility: I see in your screenshot that your DAG files are on a NAS . There is a likelihood that this it is caused by accessing NAS for some reason - maybe you should check your NAS logs ?

**Reply by potiuk** [reply]
Maybe  you also have done some maintenance/upgrade/change in your OS-level infrastructure and it coincided with airflow upgrade ?

**Comment by 4andy** [comment]
Thanks for all the tips and ideas. I have done some investigation and used py-spy as suggested above. I found that the connection to the postgres backend seems to be the function that is hanging - specifically, this connect call shows up on the airflow runner task.
![Screenshot 2023-08-10 150042](https://github.com/apache/airflow/assets/41160084/6b3ae7fb-b53c-4d81-831e-1620db06e427)

I did a python profile of a runner task processes on both healthy and non-healthy jobs. The healthy task was very quick and a search for 'psycopg2' in the profile only turned up one item and it was actually a sqlalchemy file.
![Screenshot 2023-08-10 161001](https://github.com/apache/airflow/assets/41160084/4f314468-ebad-4a8f-b88f-c68a2e6d4f90)

On the unhealthy runner task was slow and has 7 calls to the hanging connect function mentioned above. 
![Screenshot 2023-08-10 160857](https://github.com/apache/airflow/assets/41160084/d8223788-978c-4485-8896-087e93c22ea4)

That is far as I was able to work on it today. I'm guessing that the postgres connection is either not found or disconnecting or something of the sort and airflow is having to reconnect or wait for it. I didn't see any major differences on the database side of things in a quick look. I will continue debugging tomorrow but any suggestions or ideas are always welcome.

**Reply by 4andy** [reply]
I made a simple script to connect to the database with psycopg2.connect(). I can confirm that it is very slow on unhealthy servers connecting to the local database as well as remote databases. Conversely, connections from a healthy server are very fast to databases locally and to databases on unhealthy servers. As such this is definitely not an airflow issue; however, I will update this thread with the resolution when we get there if in case anyone else runs into the issue in the future.

**Reply by potiuk** [reply, marked as answer]
Thanks for being so diligent  I would love to hear what it is.

Q: Are you using PgBouncer beetween Airflow and your Postgres ? Because if you DON't then THIS is the most probable reason. Airflow opens a lot of connection to the database and that might cause the memory usage of postgres to go ballistic if you have a lot of parallel tasks and dag file processor parsing tasks. Postgres works in the way that it spawns a new process for every connection, it does not multiplex them in the way that mysql does, so well configured PGBouncer is pretty much a MUST if you want to use Postgres.

If Postgres eats all memory and start swapping out then you might indeed have a VERY slow connection establishing process.

**Comment by 4andy** [comment]
While I have not discovered the root OS item that is progressively overloading causing the db connection time to hang and be slow, PgBouncer did indeed resolve the issue. Thanks for that prompt.

**Reply by potiuk** [reply]
Glad it helped.
